File: common/src/main/java/generations/gg/generations/core/generationscore/api/player/AccountInfo.java
package generations.gg.generations.core.generationscore.api.player;

import com.cobblemon.mod.common.Cobblemon;
import com.google.gson.JsonObject;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.NotNull;

import java.math.BigDecimal;

public class AccountInfo extends PlayerDataExtension {
    public static String KEY = "account_info";
    private BigDecimal balance;

    public AccountInfo() {
        balance = BigDecimal.ZERO;
    }

    public AccountInfo(BigDecimal balance) {
        this.balance = balance;
    }

    public void setBalance(BigDecimal balance) {
        this.balance = balance;
    }

    public BigDecimal getBalance() {
        return balance;
    }

    @NotNull
    @Override
    public String name() {
        return KEY;
    }

    @NotNull
    @Override
    public JsonObject serialize() {
        var json = super.serialize();
        json.addProperty("balance", balance);
        return json;
    }

    @NotNull
    @Override
    public PlayerDataExtension deserialize(@NotNull JsonObject jsonObject) {
        return new AccountInfo(jsonObject.getAsJsonPrimitive("balance").getAsBigDecimal());
    }

    public static AccountInfo get(Player player) {
        return (AccountInfo) Cobblemon.playerData.get(player).getExtraData().computeIfAbsent(KEY, key -> new AccountInfo(BigDecimal.valueOf(1000000)));
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/api/player/AccountInfo.kt
package generations.gg.generations.core.generationscore.common.api.player

import com.google.gson.JsonObject
import generations.gg.generations.core.generationscore.common.GenerationsCore
import net.minecraft.world.entity.player.Player
import java.math.BigDecimal

class AccountInfo(var balance: BigDecimal = BigDecimal.ZERO) : PlayerDataExtension() {

    override fun name(): String {
        return KEY
    }

    override fun serialize(): JsonObject {
        val json = super.serialize()
        json.addProperty("balance", balance)
        return json
    }

    override fun deserialize(jsonObject: JsonObject): PlayerDataExtension {
        return AccountInfo(jsonObject.getAsJsonPrimitive("balance").asBigDecimal)
    }

    companion object {
        var KEY: String = "account_info"
        fun get(player: Player?): AccountInfo {
            return AccountInfo(GenerationsCore.CONFIG!!.economy.startingFunds) //Cobblemon.playerDataManager.getGenericData(player).getExtraData().computeIfAbsent(KEY, key -> new AccountInfo(GenerationsCore.CONFIG.economy.startingFunds)); TODO: rework
        }
    }
}
